cookbook/models.py

```diff
@@ -37,16 +37,73 @@ def get_model_name(model):
 
 
 class TreeManager(MP_NodeManager):
+    # model.Manager get_or_create() is not compatible with MP_Tree
     def get_or_create(self, **kwargs):
-        # model.Manager get_or_create() is not compatible with MP_Tree
         kwargs['name'] = kwargs['name'].strip()
-        q = self.filter(name__iexact=kwargs['name'], space=kwargs['space'])
-        if len(q) != 0:
-            return q[0], False
-        else:
+        try:
+            return self.get(name__iexact=kwargs['name'], space=kwargs['space']), False
+        except self.model.DoesNotExist:
             with scopes_disabled():
-                node = self.model.add_root(**kwargs)
-            return node, True
+                return self.model.add_root(**kwargs), True
```

Comment:
do we need to disable scopes here as `space` should always be present in `kwargs` and if not we want this function to fail?

Reply:
other than that i love the generic approach

Reply:
The MP_Tree needs to see the entire tree to work correctly when dealing with root objects, if it doesn't it will add a duplicate path which will cause a unique constraint error.  
We can explicitly check for the existence of 'space' and trigger a failure if it is not present.